_posts/2015-01-13-matryoshka-configurable-caching-library-for-php.md

```diff
@@ -0,0 +1,146 @@
+
+{% highlight php startinline=true %}
+$doubleExpiration = function($expiration) {
+   return $expiration * 2;
+};
+$cache = new Matryoshka\ExpirationChange($prefixedCache,
+ $doubleExpiration);
+// Results in an expiration time of 20 for "prefix-key".
+$cache->set('key', 'value', 10);
+{% endhighlight %}
+
+By using composition, caching configurations can be assembled on the fly quite easily.
+The core API is identical for all backends so the caller doesn't need to be aware of the exact configuration.
+Common configurations and base backends (like memcached connections) can be made into singletons or provided using dependency injection in your application.
+Additionally, this architecture results in very maintainable and testable code because each class has exactly one job.
+
+# Scopes
+
+Cache invalidation is hard.
+To make it easier, Matryoshka provides "cache scopes" to invalidate a group of keys at once.
+[This works][Scope.php] by prefixing all keys with a unique value that is stored in the backend using the scope name.
+
+{% highlight php startinline=true %}
+$cache = new Matryoshka\Scope($memcachedBackend, 'name');
+
+// This results in a get request to memcached for 'scope-name'
+// which results in something like '0fb4ae36'. This `set` call
+// then results in a key of '0fb4ae36-key'.
+$cache->set('key', 'value');
+$cache->set('key2', 'value2');
+$value = $cache->get('key'); // '0fb4ae36-key' => 'value'
+
+// Deleting the scope results in a new scope value e.g. 'e093f71e'.
+$cache->deleteScope();
+
+// Both of these result in a miss because the scope has a new
+// value so the keys are now prefixed with 'e093f71e-'.
+$value = $cache->get('key'); // 'e093f71e-key' => false
+$value2 = $cache->get('key2'); // 'e093f71e-key2' => false
+{% endhighlight %}
+
+We have found this to be particularly useful for scoping keys to code deploys.
+We simply put any caches that should be invalidated under the `'deploy'` scope which is deleted anytime we deploy code.
+You can also have dynamic scopes such as `"post-{$postid}"` which can be cleared anytime a specific post is modified.
+This is an implementation of [generational caching] which we make [heavy use of][caching thesis] throughout our application.
+
+Cache scopes help with cache invalidation but unfortunately don't make [naming things] any easier.
+
+# Helper Functions
+
+Matryoshka adds a few helper functions to make common operations easier.
+`getAndSet` makes populating the cache dead simple:
+
+{% highlight php startinline=true %}
+$cache = new Matryoshka\Ephemeral();
+// Calls the provided callback if the key is not found and sets
+// it in the cache before returning the value to the caller.
+$value = $cache->getAndSet('key', function() {
+   return 'value';
+});
+{% endhighlight %}
+
+Similarly, `getAndSetMultiple` makes doing multi-gets significantly easier:
+
+{% highlight php startinline=true %}
+// Array of key => id. The id's can be anything used to identify
```

Comment:
I think this [should be](https://en.wiktionary.org/wiki/ids#English) `ids`, without the apostrophe.